### PR TITLE
New reference panel: fallback to search-based code intel

### PR DESCRIPTION
**Pull request by mrnugget**
This fixes the first half of https://github.com/sourcegraph/sourcegraph/issues/30970 by renaming the `usePreciseCodeIntel` hook to `useCodeIntel` and doing a fallback to search-based code intel when querying precise-code intel yielded no results. This does **not yet** mix search-based/precise results.

* [x] Uses `LanguageSpec` definitions copy&pasted from `code-intel-extensions`
* [x] Determines which `Languagespec` to use for current document
* [x] Finds token in current document, using `LanguageSpec` to determine token boundaries and file extensions
* [x] Builds search query to query definitions
* [x] Builds search query to query references
* [x] Tag results as search-based and display that in UI
* [x] Handle clicks on definitions in mini code view correctly
* [x] Filter definitions based on language spec's filterDefinitions

## Test plan

- Tested locally by trying out search-based code intel in repositories without precise
- Confirmed that `filterDefinition` is used by printing comparisons between filtered/unfiltered
- Manually checked graphql queries
- I really want to look into how to best test this as part of #30970

### CODE REVIEW

File: client/web/src/codeintel/language-specs/comments.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import { BlockCommentStyle, CommentStyle } from './languagespec'
```
Comment (mrnugget) [inline client/web/src/codeintel/language-specs/comments.ts:1]: This and all the other files in `client/web/src/codeintel/language-specs` are copy & pasted (and slightly adapted to fit types) from https://github.com/sourcegraph/code-intel-extensions/tree/master/template/src/language-specs

File: client/web/src/codeintel/searchBased.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import { extname } from 'path'
```
Comment (mrnugget) [inline client/web/src/codeintel/searchBased.ts:1]: Everything in this file is taken from https://github.com/sourcegraph/code-intel-extensions/blob/master/template/src/search/queries.ts only the types are adjusted (so we don't need the extension API types here)

File: client/web/src/codeintel/token.ts [hunk 3]
```diff
@@ -0,0 +1,234 @@
+import { flatten } from 'lodash'
```
Comment (mrnugget) [inline client/web/src/codeintel/token.ts:1]: This file too is copy & pasted from https://github.com/sourcegraph/code-intel-extensions/blob/master/template/src/search/tokens.ts

File: client/web/src/codeintel/util/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+/**
```
Comment (mrnugget) [inline client/web/src/codeintel/util/helpers.ts:1]: Taken from https://github.com/sourcegraph/code-intel-extensions/blob/master/template/src/util/helpers.ts

File: client/web/src/codeintel/useCodeIntel.ts [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+        [fetchSearchBasedReferences, variables.path, spec.fileExts]
+    )
+
+    const fetchSearchBasedDefinitionsForToken = useCallback(
+        (searchToken: string) => {
+            const query = definitionQuery({ searchToken, path: variables.path, fileExts: spec.fileExts }).join(' ')
+            return fetchSearchBasedDefinitions({ variables: { query } })
+        },
+        [fetchSearchBasedDefinitions, variables.path, spec.fileExts]
+    )
```
Comment (mrnugget) [inline client/web/src/codeintel/useCodeIntel.ts:186]: Reviewers: these are essentially the only new things in this hook.

**Review by chrismwendt — APPROVED** [review]
LGTM. If we someday move this logic behind the GraphQL API that would let us deduplicate some of this code.

**Comment by mrnugget** [comment]
> If we someday move this logic behind the GraphQL API that would let us deduplicate some of this code.

Yup, that's the plan. "Someday" is "soon", hopefully :)

**Pull request merged**